connectors/cassandra/schema.go

```diff
@@ -37,17 +55,197 @@ func (c *Connector) TruncateScope(ctx context.Context, scope string) error {
+		return fmt.Sprintf("Missing %d tables (first is %q)", len(m.MissingTables), m.MissingTables[0])
+	}
+	return fmt.Sprintf("Missing %d columns (first is %q in table %q)", len(m.MissingColumns), m.MissingColumns[0].Column.Name, m.MissingColumns[0].Tablename)
+}
+
+// compareStructToSchema compares a dosa EntityDefinition to the gocql TableMetadata
+// There are two main cases, one that we can fix by adding some columns and all the other mismatches that we can't fix
+func compareStructToSchema(ed *dosa.EntityDefinition, md *gocql.TableMetadata, schemaErrors *RepairableSchemaMismatchError) error {
+	// Check partition keys
+	if len(ed.Key.PartitionKeys) != len(md.PartitionKey) {
+		return fmt.Errorf("Table %q primary key length mismatch (was %d should be %d)", ed.Name, len(md.PartitionKey), len(ed.Key.PartitionKeys))
+	}
+	for i, pk := range ed.Key.PartitionKeys {
+		if md.PartitionKey[i].Name != pk {
+			return fmt.Errorf("Table %q primary key mismatch (should be %q)", ed.Name, ed.Key.PartitionKeys)
+		}
+	}
+
+	// Check clustering keys
+	if len(ed.Key.ClusteringKeys) != len(md.ClusteringColumns) {
+		return fmt.Errorf("Table %q clustering key length mismatch (should be %q)", ed.Name, ed.Key.ClusteringKeys)
+	}
+	for i, ck := range ed.Key.ClusteringKeys {
+		if md.ClusteringColumns[i].Name != ck.Name {
+			return fmt.Errorf("Table %q clustering key mismatch (column %d should be %q)", ed.Name, i+1, ck.Name)
+		}
+	}
+
+	// Check each column
+	for _, col := range ed.Columns {
+		_, ok := md.Columns[col.Name]
+		if !ok {
+			schemaErrors.MissingColumns = append(schemaErrors.MissingColumns, MissingColumn{Column: *col, Tablename: ed.Name})
+		}
+		// TODO: check column type
+	}
+
+	return nil
+
+}
+
+// CheckSchema verifies that the schema passed in the registered entities matches the database
 func (c *Connector) CheckSchema(ctx context.Context, scope string, namePrefix string, ed []*dosa.EntityDefinition) (int32, error) {
-	return 0, nil
+	schemaErrors := new(RepairableSchemaMismatchError)
+
+	// TODO: unfortunately, gocql doesn't have a way to pass the context to this operation :(
+	km, err := c.Session.KeyspaceMetadata(c.KsMapper.Keyspace(scope, namePrefix))
+	if err != nil {
+		return 0, err
+	}
+	for _, ed := range ed {
+		tableMetadata, ok := km.Tables[ed.Name]
+		if !ok {
+			schemaErrors.MissingTables = append(schemaErrors.MissingTables, ed.Name)
+			continue
+		}
+		if err := compareStructToSchema(ed, tableMetadata, schemaErrors); err != nil {
+			return 0, err
+		}
+
+	}
+	if schemaErrors.HasMissing() {
+		return 0, schemaErrors
+	}
+	return int32(1), nil
 }
 
-// UpsertSchema is not implemented
+// UpsertSchema checks the schema and then updates it
 func (c *Connector) UpsertSchema(ctx context.Context, scope string, namePrefix string, ed []*dosa.EntityDefinition) (*dosa.SchemaStatus, error) {
-	panic("next diff")
+	sr, err := c.CheckSchema(ctx, scope, namePrefix, ed)
+	if repairs, ok := err.(*RepairableSchemaMismatchError); ok {
```

Comment:
This is getting a little deep. Might be worth pulling some of these things out into their own, named functions. IMHO the code would read nicely thusly:
```
if repairs, ok := err.(*RepairableSchemaMismatchError); ok {
  repairMissingTables(...)
  repairMissingColumns(...)
}
```